src/orange/BoundingBoxUtils.hh

```diff
@@ -221,6 +221,70 @@ inline bool encloses(BoundingBox<T> const& big, BoundingBox<T> const& small)
     });
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the distance to the inside of the bbox from a pos and dir.
+ *
+ * If there is no intersection, the result will be inf.
+ */
+template<class T, class U>
+inline U calc_dist_to_inside(BoundingBox<T> const& bbox,
+                             Array<U, 3> const& pos,
+                             Array<U, 3> const& dir)
+{
+    if (is_inside(bbox, pos))
+    {
+        // Starting from inside the bbox so the distance to inside is zero
+        return 0.;
+    }
+
+    // Loop over all 6 planes to find the minimum intersection
+    U min_dist = numeric_limits<U>::infinity();
+    for (auto bound : range(to_int(Bound::size_)))
+    {
+        for (auto ax : range(to_int(Axis::size_)))
+        {
+            if (dir[ax] == 0)
+            {
+                // Short circut if there is not movement in this dir
+                continue;
+            }
+
+            U dist = (bbox.point(Bound{bound})[ax] - pos[ax]) / dir[ax];
+
+            if (dist < 0)
+            {
+                // Short circut if the plane is behind us
+                continue;
+            }
+
+            // Calculate the actual intersection point
+            Array<U, 3> intersect = pos + dist * dir;
+
+            // Check that the intersection point occurs within the region
+            // bounded by the planes of the other two axes
+            bool in_bounds = true;
+            for (auto other_ax : range(to_int(Axis::size_)))
+            {
+                if (other_ax != ax
+                    && (intersect[other_ax] < bbox.lower()[other_ax]
+                        || intersect[other_ax] > bbox.upper()[other_ax]))
+                {
+                    in_bounds = false;
+                    break;
+                }
+            }
```

Comment:
I'd say to write this as an algorithm but using `range` makes it hard... using an IFFE allows you to express the logic more clearly (or you could write this as an anonymous function and pass the local arguments)
```suggestion
            bool in_bounds = [&] {
                for (auto other_ax : range(to_int(Axis::size_)))
                {
                    if (other_ax == ax)
                        continue;

                    if (intersect[other_ax] < bbox.lower()[other_ax]
                        || intersect[other_ax] > bbox.upper()[other_ax])
                    {
                        return false;
                    }
                }
                return true;
            }();
```
If we wanted to be robust against NaNs, we'd have to write:
```c++
                    if (!(intersect[other_ax] >= bbox.lower()[other_ax]
                        && intersect[other_ax] <= bbox.upper()[other_ax]))
```

Reply:
And actually you could probably move this function outside of the loop so that the intersect is an argument. And you should also consider evaluating `pos + dist * dir` for one element at a time, since you never need them all simultaneously (and in fact you *always* exclude one of them), to reduce register pressure.